### ISSUE TITLE
[feature] self upgrading release binary

### ISSUE THREAD

**Issue by kskarthik** [passbolt/go-passbolt-cli#50]
The idea is to have similar feature to `deno upgrade` & `bun upgrade` in which the user can upgrade the existing binary using a cli flag.

This will be a useful feature for many users who do not use distributions which always do not ship with latest versions of the cli

I wish to work on this patch if we can agree on implementing such feature.

**Comment by kskarthik** [comment]
i have initially found this lib which can do the heavy lifing https://github.com/creativeprojects/go-selfupdate

**Comment by speatzle** [comment]
I don't see how that library can update additional files like man pages or shell completions (maybe i missed it?). Also if anything goes wrong and can't start it anymore there is no possibility to downgrade again. I generally dislike the Windows Style Self updating.

I don't see adding custom repositories as such a big deal (you also have to do that if you are installing the passbolt server). go-passbolt-cli Simply is not big enough to be included in distros.

I tried to avoid having to add custom repos by looking into packaging the CLI as a Flatpak instead but their CLI Support has "Issues" we would have to work around https://github.com/flatpak/flatpak/issues/1188 .

**Comment by kskarthik** [comment]
> I don't see how that library can update additional files like man pages or shell completions (maybe i missed it?). Also if anything goes wrong and can't start it anymore there is no possibility to downgrade again. I generally dislike the Windows Style Self updating.

Yes, The lib i shared does not handle man pages & shell completions. 

> I tried to avoid having to add custom repos by looking into packaging the CLI as a Flatpak instead but their CLI Support has "Issues" we would have to work around https://github.com/flatpak/flatpak/issues/1188 .

yes, flatpaks are not ideal for CLIs. We have to add an alias & no support for man pages AFAIK

**Comment by kskarthik** [comment]
I have also packaged for opensuse tumbleweed https://build.opensuse.org/request/show/1181068

It takes a few days to land in the main repositories!

**Comment by speatzle** [comment]
> I have also packaged for opensuse tumbleweed https://build.opensuse.org/request/show/1181068

Cool, i will add that to the Readme once it has been released

**Comment by kskarthik** [comment]
Sure, i will let you know

**Comment by kskarthik** [comment]
@speatzle Now accepted to tumbleweed! https://build.opensuse.org/package/show/openSUSE:Factory/go-passbolt-cli

**Comment by kskarthik** [comment]
It might also reflect here https://repology.org/project/go:passbolt-cli/versions shortly

**Comment by speatzle** [comment]
Cool, i will give it a spin this weekend.

**Comment by kskarthik** [comment]
btw is this project still community driven or made official ?

**Comment by speatzle** [comment]
Yes, this Project is still Community Driven.

**Comment by speatzle** [comment]
Hey @kskarthik 
Sorry for the delay. I finally managed to setup a tumbleweed VM (the few attempts ended in emergency mode due to broken virtio drivers...) and Test the Package.

It Seems to work well but both shell completions and man pages seem to be missing making it hard to use, could you add those to the package?

You can see how to generate them here: https://github.com/passbolt/go-passbolt-cli/blob/main/.github/workflows/.release.yml

**Comment by kskarthik** [comment]
Sure, I will

**Comment by kskarthik** [comment]
just added shell completions & man page support https://build.opensuse.org/package/rdiff/security:privacy/go-passbolt-cli?opackage=go-passbolt-cli&oproject=openSUSE%3AFactory&rev=4

Will reflect in tumbleweed in a few days

**Comment by speatzle** [comment]
I upgraded the package and the man pages work now, it also installed go-passbolt-cli-bash-completions during the upgrade but i can't get the completions to work. Do i need to enable them manually?
![image](https://github.com/user-attachments/assets/fb9733c1-d993-4cc4-b9e8-b72dc5c12d02)

**Comment by kskarthik** [comment]
do you have the bash completions package installed ?

**Comment by speatzle** [comment]
Yes, it was installed automatically:
```
slt@localhost:~> zypper search passbolt
Loading repository data...
Reading installed packages...

S  | Name                            | Summary                              | Type
---+---------------------------------+--------------------------------------+--------
i+ | go-passbolt-cli                 | A CLI tool to interact with Passbolt | package
i  | go-passbolt-cli-bash-completion | Bash Completion for go-passbolt-cli  | package
   | go-passbolt-cli-fish-completion | Fish Completion for go-passbolt-cli  | package
   | go-passbolt-cli-zsh-completion  | ZSH Completion for go-passbolt-cli   | package
slt@localhost:~> zypper info bash-completion
Loading repository data...
Reading installed packages...


Information for package bash-completion:
----------------------------------------
Repository     : Main Repository (OSS)
Name           : bash-completion
Version        : 2.12.0-2.2
Arch           : noarch
Vendor         : openSUSE
Installed Size : 1003.5 KiB
Installed      : Yes (automatically)
Status         : up-to-date
Source package : bash-completion-2.12.0-2.2.src
Upstream URL   : https://github.com/scop/bash-completion/
Summary        : Programmable Completion for Bash
Description    : 




    bash-completion is a collection of shell functions that take advantage
    of the programmable completion feature of Bash 2.04 and later.

slt@localhost:~> zypper info go-passbolt-cli-bash-completion
Loading repository data...
Reading installed packages...


Information for package go-passbolt-cli-bash-completion:
--------------------------------------------------------
Repository     : Main Repository (OSS)
Name           : go-passbolt-cli-bash-completion
Version        : 0.3.1-2.1
Arch           : noarch
Vendor         : openSUSE
Installed Size : 69.1 KiB
Installed      : Yes (automatically)
Status         : up-to-date
Source package : go-passbolt-cli-0.3.1-2.1.src
Upstream URL   : https://github.com/passbolt/go-passbolt-cli
Summary        : Bash Completion for go-passbolt-cli
Description    : 
    The official bash completion script for go-passbolt-cli, generated during the build.

```

**Comment by kskarthik** [comment]
ok, https://build.opensuse.org/projects/openSUSE:Factory/packages/go-passbolt-cli/files/go-passbolt-cli.spec?expand=1 L94 is where the bash completions go

**Comment by kskarthik** [comment]
i can see the bash completions listed. do you find any anomaly?

```
usr/share/man/man1/passbolt.1.gz
/usr/share/bash-completion/completions/go-passbolt-cli

**Comment by kskarthik** [comment]
should the completions match the binary name ?

**Comment by speatzle** [comment]
Moving to file to match the binary name fixed it for me
```
slt@localhost:~> sudo mv /usr/share/bash-completion/completions/go-passbolt-cli  /usr/share/bash-completion/completions/passbolt
[sudo] password for root: 
slt@localhost:~> bash
slt@localhost:~> passbolt 
configure  create     delete     export     get        help       list       move       share      update     verify     
slt@localhost:~> passbolt list 
folder    group     resource  user      
slt@localhost:~> passbolt list resource 

```

**Comment by kskarthik** [comment]
ok, i will correct it upstream too
